waspc/src/Wasp/Generator/Valid/PackageJson.hs

```diff
@@ -12,163 +18,180 @@ import Wasp.Generator.DepVersions (prismaVersion, typescriptVersion)
+  whenS (inOppositeDepList notPresentValidator pkgJson) $
+    inCorrectDepList correctVersionValidator pkgJson
   where
-    maybePackageJsonDepedency :: Maybe P.PackageVersion
-    maybePackageJsonDepedency = case requirement of
-      RequiredRuntime -> M.lookup packageName $ P.dependencies packageJson
-      RequiredDevelopment -> M.lookup packageName $ P.devDependencies packageJson
-      Optional ->
-        M.lookup packageName (P.dependencies packageJson)
-          <|> M.lookup packageName (P.devDependencies packageJson)
-
-    isInWrongLocation :: PackageRequirement -> Bool
-    isInWrongLocation = \case
-      RequiredRuntime -> M.member packageName (P.devDependencies packageJson)
-      RequiredDevelopment -> M.member packageName (P.dependencies packageJson)
-      Optional -> False
-
-    getMissingPackageError :: PackageRequirement -> [GeneratorError]
-    getMissingPackageError = \case
-      RequiredRuntime -> [missingRequiredPackageErrorMessage]
-      RequiredDevelopment -> [missingRequiredPackageErrorMessage]
-      Optional -> []
-
-    incorrectPackageVersionErrorMessage :: GeneratorError
-    incorrectPackageVersionErrorMessage =
-      GenericGeneratorError $
-        unwords
-          [ "Wasp requires package",
-            show packageName,
-            "to be version",
-            show expectedPackageVersion,
-            "in package.json."
-          ]
-
-    missingRequiredPackageErrorMessage :: GeneratorError
-    missingRequiredPackageErrorMessage =
-      GenericGeneratorError $
-        unwords
-          [ "Wasp requires package",
-            show packageName,
-            "with version",
-            show expectedPackageVersion,
-            "to be in",
-            show $ getExpectedPackageJsonDependencyKey requirement,
-            "in package.json."
-          ]
-
-    wrongDependencyTypeErrorMessage :: GeneratorError
-    wrongDependencyTypeErrorMessage =
-      GenericGeneratorError $
-        unwords $
-          [ "Wasp requires package",
-            show packageName,
-            "to be in",
-            show $ getExpectedPackageJsonDependencyKey requirement
-          ]
-            ++ ( case getOppositePackageJsonDependencyKey requirement of
-                   Just oppositeKey -> ["and not in", show oppositeKey]
-                   Nothing -> []
-               )
-            ++ ["in package.json."]
-
-getExpectedPackageJsonDependencyKey :: PackageRequirement -> String
-getExpectedPackageJsonDependencyKey = \case
-  RequiredRuntime -> "dependencies"
-  RequiredDevelopment -> "devDependencies"
-  Optional -> "dependencies or devDependencies"
-
-getOppositePackageJsonDependencyKey :: PackageRequirement -> Maybe String
-getOppositePackageJsonDependencyKey = \case
-  RequiredRuntime -> Just "devDependencies"
-  RequiredDevelopment -> Just "dependencies"
-  Optional -> Nothing
+    notPresentValidator :: V.Validator (Maybe P.PackageVersion) Bool
```

Comment:
Curious, why is this one Bool and not ()? It doesn't seem like it will ever return False, it will be either error or success, so () should be enough to model that?